BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,162 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/S3SaveRestoreLib.h>
+#include <Library/IoLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Library/BoardInitLib.h>
+#include <Guid/SmmInformationGuid.h>
+
+
+/**
+  This function uploads information in TSEG area
+  designated for S3 save/restore purpose.
+
+  @param    DataPtr       Address of the structure to be copied to TSEG
+
+  @retval   None
+**/
+VOID
+EFIAPI
+SetS3Info (
+  IN  VOID    *DataPtr
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  SmmSize = LdrSmmInfo.SmmSize;
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature != 0x00000000) {
+      Ptr += ((BL_PLD_COMM_HDR *)Ptr)->TotalSize;
+      continue;
+    }
+    break;
+  }
+
+  if ( Ptr + ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize > SmmBase + SmmSize ) {
+    DEBUG ((DEBUG_ERROR, "Overflowing SMM region\n"));
+    return;
+  }
+  CopyMem ((VOID *)Ptr, DataPtr, ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize);
+}
+
+
+/**
+  Get pointer to a specific struct area in TSEG
+
+  @param      Signature         Signature of the struct to be retreived
+
+  @retval     NULL              Struct with signature not found
+  @retval     Ptr               Pointer to struct with signature
+**/
+VOID *
+EFIAPI
+GetS3Info (
+  IN  UINT32  Signature
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  SmmSize = LdrSmmInfo.SmmSize;
+
+  if ( ((BL_PLD_COMM_HDR *)SmmBase)->Signature != SMM_BASE_INFO_SIG) {
+    DEBUG((DEBUG_ERROR, "Invalid communication area\n"));
+    return NULL;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature != Signature) {
+      Ptr += ((BL_PLD_COMM_HDR *)Ptr)->TotalSize;
+      continue;
+    }
+    break;
+  }
+
+  return (VOID *) Ptr;
+}
+
+
+/**
+  This function restores the states of the registers that were
+  set to be saved by the bootloader in the normal boot path.
+  After the payload has updated the 'Val' field of the reg info
+  in the normal boot path, we simply OR the updated 'Val'
+  with the existing vale of the register in the S3 resume boot path.
+  This function is only called in the S3 resume path.
+
+  @param    SmmBaseInfo   TSEG Base
+    
+  @retval   None
+
+**/
+VOID
+EFIAPI
+RestoreS3RegInfo (
+  IN  S3_SAVE_REG   *S3SaveReg
```

Comment:
Parameters in function comment doesn't match actual name.

Reply:
ok, will change.